docs/proposals/Certificate Revocation Lists.md

```diff
@@ -0,0 +1,213 @@
+
+#### Save CRL to Cache
+
+```
+// Set stores the CRL bundle in the file system
+// Check closest expired date and set to `CRLCacheProvider`.
+// Save to temp file and avoid concurrency issue with atomic write operation
+// `rename()` is atomic on UNIX-like platforms
+
+
+// notation-go FScache
+
+type fileCacheContent struct {
+	// BaseCRL is the ASN.1 encoded base CRL
+	BaseCRL []byte `json:"baseCRL"`
+
+	// DeltaCRL is the ASN.1 encoded delta CRL
+	DeltaCRL []byte `json:"deltaCRL,omitempty"`
+}
+
+// This cache builds on top of the UNIX file system to leverage the file system's
+// atomic operations. The `rename` and `remove` operations will unlink the old
+// file but keep the inode and file descriptor for existing processes to access
+// the file. The old inode will be dereferenced when all processes close the old
+// file descriptor. Additionally, the operations are proven to be atomic on
+// UNIX-like platforms, so there is no need to handle file locking.
+//
+// NOTE: For Windows, the `open`, `rename` and `remove` operations need file
+// locking to ensure atomicity. The current implementation does not handle
+// file locking, so the concurrent write from multiple processes may be failed.
+// Please do not use this cache in a multi-process environment on Windows.
+```
+
+### Provide CRL Cache
+
+#### Get Cache from Provider
+
+```
+// Get retrieves the CRL bundle from the file system
+//
+// If the key does not exist, return ErrNotFound
+// If the CRL is expired, return ErrCacheMiss
+
+
+// Policy that uses custom verification level to relax the strict verification.
+// It logs expiry and skips the revocation check.
+"name": "use-expired-blobs",
+    "signatureVerification": {
+        "level" : "strict",
+        "override" : {
+            "expiry" : "log",
+            "revocation" : "skip"
+        }
+    },
+
+```
+Reference: https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#signatureverification-details:~:text=Notary%20Project%20defines,scope%20(*).
+
+#### Refresh Cache
+
+Cache Data Structure: Store data along with expiration timestamps.
+
+Monitor Scheduler: Use a scheduler (e.g., time.Ticker) to check the cache at regular intervals.
+
+Concurrency: Use synchronization mechanisms like mutexes for thread-safe access to shared data.
+
+Expiration Handling: When setting cache, check closest expired date that set to `CRLCacheProvider`. Compare the current time with the cache item's expiration. If expired, trigger the fetch process to update the data.
+
+Error Handling and Retries: Implement error handling with retry logic in case of failed refresh operations.
+
+Use synchronization primitives like mutexes to ensure thread safety during cache updates.
+
+# More details
```

Comment:
Please include dev work breakdown , the cost should factor in unit test/ perf test / e2e test. https://github.com/ratify-project/ratify/blob/dev/docs/design/Cosign%20Upgrade%202024.md#dev-work-items